### PR TITLE
style: Move font-size outside of mako

### LINKED ISSUE #19111: Move font-size outside of mako.
I'll take this one as a sub-issue of #19015

### PR DESCRIPTION
**Pull request by CYBAI**
This is a sub-PR of #19015 
r? emilio 

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #19111  (github issue number if applicable).
- [x] These changes do not require tests

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19119)

### CODE REVIEW

File: components/style/values/specified/font.rs [hunk 1]
```diff
@@ -466,6 +466,110 @@ impl FontSize {
             None
         }
     }
+
+    #[inline]
+    /// Get initial value for specified font size.
+    pub fn get_initial_specified_value() -> Self {
```
Comment (emilio) [inline components/style/values/specified/font.rs]: Let's call this method `fn medium()` instead.

File: components/style/values/specified/font.rs [hunk 2]
```diff
@@ -466,6 +466,110 @@ impl FontSize {
...
+        FontSize::Keyword(computed::KeywordInfo::medium())
+    }
+
+    /// Parses a font-size, with quirks.
+    pub fn parse_quirky<'i, 't>(context: &ParserContext,
```
Comment (emilio) [inline components/style/values/specified/font.rs]: Let's fix the indentation here:

```rust
pub fn parse_quirky<'i, 't>(
    context: &ParserContext,
    input: &mut Parser<'i, 't>,
    allow_quirks: AllowQuirks,
) -> Result<Self, ParseError<'i>> {
```

File: components/style/values/specified/font.rs [hunk 3]
```diff
@@ -466,6 +466,110 @@ impl FontSize {
...
+
+        try_match_ident_ignore_ascii_case! { input,
+            "smaller" => Ok(FontSize::Smaller),
+            "larger" => Ok(FontSize::Larger),
+        }
+    }
+
+    #[allow(unused_mut)]
+    /// Cascade `font-size` with specified value
+    pub fn cascade_specified_font_size(context: &mut Context,
```
Comment (emilio) [inline components/style/values/specified/font.rs]: ditto re. the indentation.

File: components/style/values/specified/font.rs [hunk 4]
```diff
@@ -466,6 +466,110 @@ impl FontSize {
...
+                specified_value.to_computed_value_against(context, FontBaseSize::Custom(Au::from(parent)));
+            context.builder
+                   .mutate_font()
+                   .apply_unconstrained_font_size(new_unconstrained.size);
+        }
+    }
+
+    /// FIXME(emilio): This is very complex. Also, it should move to
+    /// StyleBuilder.
+    pub fn cascade_inherit_font_size(context: &mut Context) {
```
Comment (emilio) [inline components/style/values/specified/font.rs]: These two should probably be in `computed::FontSize`, but it's not a huge deal.

File: components/style/values/computed/font.rs [hunk 5]
```diff
@@ -139,6 +139,15 @@ impl FontSize {
     pub fn size(self) -> Au {
         self.size.into()
     }
+
+    #[inline]
+    /// Get default value of font size.
+    pub fn get_initial_value() -> Self {
```
Comment (emilio) [inline components/style/values/computed/font.rs]: Again, let's call this `medium`, since it's the default value.

**Review by emilio — COMMENTED** [review]
Looks great, as usual, thanks a lot @CYBAI!

File: components/style/values/specified/font.rs (hunk 4 above)
Comment (CYBAI) [reply components/style/values/specified/font.rs]: Updated! I was originally we need to cascade when specified so I put it in specified module.
However, I think there's already a `cascade_specified_font_size` so I think you're right that I need to move these two into `computed` module.
Please review, thanks! 😄

**Comment by emilio** [comment]
@bors-servo r+

Thanks a lot!

**Comment by emilio** [comment]
That failure seems related, I'm going to retry just in case it isn't, but otherwise it should be investigated.

@bors-servo retry

**Comment by CYBAI** [comment]
Still failed.
It seems that it's related to quirk mode.
How could I trace it or test it in local ?

**Comment by emilio** [comment]
You should be able to test it with `./mach test /_mozilla/mozilla/unitless-length.html`

**Comment by emilio** [comment]
@bors-servo r+

#19136 landed, so this should be fine now. Thanks a lot @CYBAI!

**Comment by jdm** [comment]
Still the same:
```
  ▶ Unexpected subtest result in /quirks-mode/unitless-length.html:
  │ FAIL [expected PASS] Excluded property font:normal normal normal 40 sans-serif
  │   → assert_equals: quirk was supported expected "16px" but got "40px"
  │ 
  │ onload/</<@http://web-platform.test:8000/quirks-mode/unitless-length.html:209:17
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1489:20
  │ test@http://web-platform.test:8000/resources/testharness.js:511:9
  │ onload/<@http://web-platform.test:8000/quirks-mode/unitless-length.html:207:13
  └ onload@http://web-platform.test:8000/quirks-mode/unitless-length.html:206:9

  ▶ Unexpected subtest result in /_mozilla/mozilla/unitless-length.html:
  │ FAIL [expected PASS] Excluded property font:normal normal normal 40 sans-serif
  │   → assert_equals: quirk was supported expected "16px" but got "40px"
  │ 
  │ onload/</<@http://web-platform.test:8000/_mozilla/mozilla/unitless-length.html:209:17
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1489:20
  │ test@http://web-platform.test:8000/resources/testharness.js:511:9
  │ onload/<@http://web-platform.test:8000/_mozilla/mozilla/unitless-length.html:207:13
  └ onload@http://web-platform.test:8000/_mozilla/mozilla/unitless-length.html:206:9
```

**Comment by emilio** [comment]
Err, I'm stupid, sorry for overlooking this.

FontSize::parse right now is:

parse_quirky(..., AllowQuirks::No)

But it was:

parse_quirky(..., AllowQuirks::Yes)

So, that explains the bug, but looks fishy to me.

**Comment by emilio** [comment]
So `font_size::parse` right now doesn't accept quirks, and the reason it doesn't, AIUI, is because it'd be ambiguous with line-height, font-weight, and other properties... I'm trying to see whether there's a spec here saying what to do or not.

**Comment by emilio** [comment]
So browsers seem to agree... Here's my proposed fix:

```patch
From 9deb8d9b254f4b2e3fccec3681a78f9b1a4e3b69 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Emilio=20Cobos=20=C3=81lvarez?= <emilio@crisal.io>
Date: Wed, 8 Nov 2017 00:23:18 +0100
Subject: [PATCH] fixup! style: Move font-size outside of mako
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Emilio Cobos Álvarez <emilio@crisal.io>
---
 components/style/properties/shorthand/font.mako.rs | 32 +++++++++++++++-------
 1 file changed, 22 insertions(+), 10 deletions(-)

diff --git a/components/style/properties/shorthand/font.mako.rs b/components/style/properties/shorthand/font.mako.rs
index b06f03a74e..cd70393a84 100644
--- a/components/style/properties/shorthand/font.mako.rs
+++ b/components/style/properties/shorthand/font.mako.rs
@@ -20,10 +20,11 @@
                     spec="https://drafts.csswg.org/css-fonts-3/#propdef-font">
     use parser::Parse;
     use properties::longhands::{font_family, font_style, font_weight, font_stretch};
-    use properties::longhands::{font_size, font_variant_caps};
+    use properties::longhands::font_variant_caps;
     #[cfg(feature = "gecko")]
     use properties::longhands::system_font::SystemFont;
     use values::specified::text::LineHeight;
+    use values::specified::FontSize;
 
     <%
         gecko_sub_properties = "kerning language_override size_adjust \
@@ -89,27 +90,38 @@
                     continue
                 }
             }
-            size = Some(font_size::parse(context, input)?);
+            size = Some(FontSize::parse(context, input)?);
             break
         }
+
+        let size = match size {
+            Some(s) => s,
+            None => {
+                return Err(input.new_custom_error(StyleParseErrorKind::UnspecifiedError))
+            }
+        };
+
+        let line_height = if input.try(|input| input.expect_delim('/')).is_ok() {
+            Some(LineHeight::parse(context, input)?)
+        } else {
+            None
+        };
+
         #[inline]
         fn count<T>(opt: &Option<T>) -> u8 {
             if opt.is_some() { 1 } else { 0 }
         }
-        if size.is_none() ||
-           (count(&style) + count(&weight) + count(&variant_caps) + count(&stretch) + nb_normals) > 4 {
+
+        if (count(&style) + count(&weight) + count(&variant_caps) + count(&stretch) + nb_normals) > 4 {
             return Err(input.new_custom_error(StyleParseErrorKind::UnspecifiedError))
         }
-        let line_height = if input.try(|input| input.expect_delim('/')).is_ok() {
-            Some(LineHeight::parse(context, input)?)
-        } else {
-            None
-        };
+
         let family = FontFamily::parse(input)?;
         Ok(expanded! {
-            % for name in "style weight stretch size variant_caps".split():
+            % for name in "style weight stretch variant_caps".split():
                 font_${name}: unwrap_or_initial!(font_${name}, ${name}),
             % endfor
+            font_size: size,
             line_height: line_height.unwrap_or(LineHeight::normal()),
             font_family: family,
             % if product == "gecko":
-- 
2.15.0
```

**Comment by CYBAI** [comment]
@emilio I'll try to fix it with what you proposed.
Besides, could you provide which spec you check to make me understand it more clear?
Thanks.

**Comment by CYBAI** [comment]
Got it! I think that's why you make the `match size` do the `None` check earlier?

File: components/style/properties/shorthand/font.mako.rs [hunk 6]
```diff
@@ -50,7 +51,11 @@
             if let Ok(sys) = input.try(SystemFont::parse) {
                 return Ok(expanded! {
                      % for name in SYSTEM_FONT_LONGHANDS:
-                         ${name}: ${name}::SpecifiedValue::system_font(sys),
+                         % if name == "font_size":
+                             ${name}: FontSize::system_font(sys),
+                         % else:
+                             ${name}: ${name}::SpecifiedValue::system_font(sys),
+                         % endif
```
Comment (CYBAI) [inline components/style/properties/shorthand/font.mako.rs]: I saw it got [error](https://travis-ci.org/servo/servo/jobs/298879594) from this line in Travis.
Due to unfamiliar with the `SYSTEM_FONT_LONGHANDS`, I just tried to change it with this ugly solution.
If there's any better solution, please tell me and I'll update it.

Comment (emilio) [reply components/style/properties/shorthand/font.mako.rs]: Yeah, this is fine, I can't think of anything better right now. This system_font thing is quite messy too...

**Review by emilio — APPROVED** [review]

**Comment by emilio** [comment]
@bors-servo retry

That crash can't be caused by this PR, though it looks worrying, cc @jdm

```
  ▶ CRASH [expected OK] /css/cssom/interfaces.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 17.3.0-devel
  │ Stack trace for thread "ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(NonZero(NonZero(1))) }"
  │ stack backtrace:
  │    0:     0x7fbb8fba8d3c - backtrace::backtrace::trace::h3049d3b23f5565ad
  │    1:     0x7fbb8fba8d72 - backtrace::capture::Backtrace::new::h48230b597774155d
  │    2:     0x7fbb8f91104d - servo::install_crash_handler::handler::h1891983b809387e0
  │    3:     0x7fbb919bf2a3 - AsmJSFaultHandler
  │                         at /home/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/424067c/mozjs/js/src/asmjs/WasmSignalHandlers.cpp:1171
  │    4:     0x7fbb8d81e32f - <unknown>
  │    5:     0x7fbb849e9026 - <unknown>
  │ Stack trace for thread "ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(NonZero(NonZero(1))) }"
  │ stack backtrace:
  │    0:     0x7fbb8fba8d3c - backtrace::backtrace::trace::h3049d3b23f5565ad
  │    1:     0x7fbb8fba8d72 - backtrace::capture::Backtrace::new::h48230b597774155d
  │    2:     0x7fbb8f91104d - servo::install_crash_handler::handler::h1891983b809387e0
  │    3:     0x7fbb8d05dcaf - <unknown>
  │    4:     0x7fbb8f9110c0 - servo::install_crash_handler::handler::h1891983b809387e0
  │    5:     0x7fbb919bf2a3 - AsmJSFaultHandler
  │                         at /home/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/424067c/mozjs/js/src/asmjs/WasmSignalHandlers.cpp:1171
  │    6:     0x7fbb8d81e32f - <unknown>
  └    7:     0x7fbb849e9026 - <unknown>
```

File: components/style/properties/shorthand/font.mako.rs (hunk 6 above)
Comment (CYBAI) [reply components/style/properties/shorthand/font.mako.rs]: Thanks for review!

**Pull request merged**